dace/frontend/python/astutils.py

```diff
@@ -480,25 +481,27 @@ class ExtNodeVisitor(ast.NodeVisitor):
         top-level expressions in bodies in order to discern DaCe statements 
         from others. """
     def visit_TopLevel(self, node):
-        clsname = type(node).__name__
-        if getattr(self, "visit_TopLevel" + clsname, False):
-            getattr(self, "visit_TopLevel" + clsname)(node)
+        visitor_name = "visit_TopLevel" + type(node).__name__
+        if hasattr(self, visitor_name):
+            visitor = getattr(self, visitor_name)
+            return visitor(node)
         else:
-            self.visit(node)
+            return self.visit(node)
 
     def generic_visit(self, node):
         for field, old_value in ast.iter_fields(node):
             if isinstance(old_value, list):
                 for value in old_value:
                     if isinstance(value, ast.AST):
                         if (field == 'body' or field == 'orelse'):
-                            clsname = type(value).__name__
-                            if getattr(self, "visit_TopLevel" + clsname, False):
-                                getattr(self, "visit_TopLevel" + clsname)(value)
+                            visitor_name = "visit_TopLevel" + type(value).__name__
+                            if hasattr(self, visitor_name):
+                                visitor = getattr(self, visitor_name)
+                                return visitor(value)
                             else:
-                                self.visit(value)
+                                return self.visit(value)
                         else:
-                            self.visit(value)
+                            return self.visit(value)
```

Comment:
I don't think we should return here, otherwise we cause an early exit of the loop and the remaining elements of the `old_value` list won't be visited.

This is an `ast.NodeVisitor`, so modifications to the AST are not intended. Not sure what `ast.NodeVisitor` does when it goes through a list. I assume it won't `return` either.

Reply:
You are right it is inconsistent. I do not know why I did it, I think the final return irritated my.
What is more interesting is, that _why_ all tests have passed?

Reply:
Yes, I was surprised by this too. Unfortunately, this is dead-code testing wise:
https://app.codecov.io/gh/spcl/dace/blob/master/dace%2Ffrontend%2Fpython%2Fastutils.py#L492

The whole `ExtNodeVisitor` and `ExtNodeTransformer` classes seem a bit strange to me. I don't see how a simple `visit_Expr` basically accomplishes the same with the normal `ast.NodeVisitor` and `ast.NodeTransformer` classes.
I suspect that in older Python versions things were different. I don't think it's worth spending more time into this.

Reply:
I agree.